lib/jit_preloader/active_record/base.rb

```diff
@@ -41,11 +41,22 @@ def has_many_aggregate(assoc, name, aggregate, field, default: 0)
             association_scope = klass.all.merge(association(assoc).scope).unscope(where: aggregate_association.foreign_key)
             association_scope = association_scope.instance_exec(&reflection.scope).reorder(nil) if reflection.scope
 
-            conditions[aggregate_association.table_name] = { aggregate_association.foreign_key => primary_ids }
+            # If the query uses an alias for the association, use that instead of the table name
+            table_alias_name = association_scope.references_values.first || aggregate_association.table_name
```

Comment:
I'm not convinced that `association_scope.references_values.first` be the value you are looking for. 

If the `association_scope` happens to have multiple references this may be incorrect.

```
association_scope.references(:contact).references_values
# => ["contacts_employees", "contact"]
```

How can we ensure that we are always referencing the correct table alias (in particular if there are multiple references)

Reply:
Not perfect either, but we can also get some information form the `joins_values`

```
association_scope.joins_values.first.left.right
# => "companies_company_employees"
```

Though this doesn't entirely work with the specs. Just post here for reference while I was digging.

Reply:
I tried to simulate a case where there would be multiple aliases for a single table. For reference, I did the following:
- Added references 2 more levels deep for the employee relations to the `ContactBook` model
```
  has_many :company_employees_of_employees, through: :company_employees, source: :employees
  has_many :peons, through: :company_employees_of_employees, source: :employees
```
- Added an aggregation method for peons to the `ContactBook` model
```
  has_many_aggregate :peons, :count, :count, "*"
```
- Added data to the nested child through another child context (https://github.com/clio/jit_preloader/blob/93ccddd3fd42c81617b6ccbc671e63504cf589e9/spec/lib/jit_preloader/preloader_spec.rb#L79)
```
      let!(:company_employee_of_employee1) { Employee.create(name: "Company Employee of Employee1", contact: company_employee1) }
      let!(:company_employee_of_employee2) { Employee.create(name: "Company Employee of Employee2", contact: company_employee2) }
      let!(:peon1) { Employee.create(name: "Company Employee of Employee1", contact: company_employee_of_employee1) }
      let!(:peon2) { Employee.create(name: "Company Employee of Employee2", contact: company_employee_of_employee2) }
```
- Modified the spec in that context to use `peons_count` instead of `company_employees_count`
- Added a debugger before https://github.com/clio/jit_preloader/blob/fe07e37e6ae60ba34178be2796123739699f5622/lib/jit_preloader/active_record/base.rb#L46 and ran the spec

The result gave me the following:
```
(byebug) association_scope.references_values
["companies_peons"]
```

The query that it executed was:
```sql
SELECT COUNT(*) AS count_all, companies_peons.contact_book_id AS companies_peons_contact_book_id
FROM "contacts"
INNER JOIN "contacts" "company_employees_of_employees_peons" ON "contacts"."contact_id" = "company_employees_of_employees_peons"."id"
INNER JOIN "contacts" "company_employees_peons" ON "company_employees_of_employees_peons"."contact_id" = "company_employees_peons"."id"
INNER JOIN "contacts" "companies_peons" ON "company_employees_peons"."contact_id" = "companies_peons"."id"
WHERE "contacts"."type" IN ('Employee') AND "companies_peons"."contact_book_id" = ? AND "companies_peons"."type" = ?
GROUP BY companies_peons.contact_book_id
```

I'm not sure if this is conclusive enough to demonstrate that the current method works. I'll continue to dig, but I'm not sure what type of query would result in multiple results for `references_values`.

Reply:
@mctaylorpants and I stepped through this as well. Unfortunately Rails doesn't seem to provide a better way to detect this exact scenario but we also confirmed the results make sense and our test suite covers what we support.

We'd like to suggest a slight code change to better communicate the intention:

```ruby
table_alias_name = association_scope.references_values.first 
name = table_alias_name || aggregate_association.table_name

conditions[name] = { aggregate_association.foreign_key => primary_ids }
```